File: frontend/services/employeeService.ts
import api from "@/lib/api";
import { EmployeeResponse, Employee } from "@/types/employee";

// ===============================
// Get Employees
// ===============================
export async function getEmployees(
  page = 1,
  limit = 20,
  search = "",
  department_id?: number,
  project_id?: number,
  status?: string
): Promise<EmployeeResponse> {
  const params = new URLSearchParams();

  params.append("page", page.toString());
  params.append("limit", limit.toString());

  if (search) {
    params.append("search", search);
  }

  if (department_id) {
    params.append("department_id", department_id.toString());
  }

  if (project_id) {
    params.append("project_id", project_id.toString());
  }

  if (status) {
    params.append("status", status);
  }

  const response = await api.get(`/employees?${params.toString()}`);

  return response.data;
}

// ===============================
// Create Employee
// ===============================
export async function createEmployee(
  employee: Partial<Employee>
): Promise<Employee> {
  const response = await api.post("/employees/", employee);

  return response.data;
}

// ===============================
// Update Employee
// ===============================
export async function updateEmployee(
  id: number,
  employee: Partial<Employee>
): Promise<Employee> {
  const response = await api.put(`/employees/${id}`, employee);

  return response.data;
}

// ===============================
// Delete Employee
// ===============================
export async function deleteEmployee(
  id: number
): Promise<void> {
  await api.delete(`/employees/${id}`);
}

File: backend/app/services/employee_service.py
from sqlalchemy.orm import Session
from sqlalchemy import or_

from app.models.employee import Employee
from app.schemas.employee import EmployeeCreate, EmployeeUpdate


def get_all_employees(
    db: Session,
    page: int = 1,
    limit: int = 20,
    search: str = None,
    department_id: int = None,
    project_id: int = None,
    status: str = None,
):
    query = db.query(Employee)

    if search:
        query = query.filter(
            or_(
                Employee.name.ilike(f"%{search}%"),
                Employee.email.ilike(f"%{search}%"),
                Employee.employee_code.ilike(f"%{search}%"),
            )
        )

    if department_id:
        query = query.filter(Employee.department_id == department_id)

    if project_id:
        query = query.filter(Employee.project_id == project_id)

    if status:
        query = query.filter(Employee.status == status)

    total = query.count()

    employees = (
        query.order_by(Employee.id)
        .offset((page - 1) * limit)
        .limit(limit)
        .all()
    )

    return {
        "employees": employees,
        "total": total,
        "page": page,
        "limit": limit,
        "pages": (total + limit - 1) // limit,
    }


def get_employee(db: Session, employee_id: int):
    return (
        db.query(Employee)
        .filter(Employee.id == employee_id)
        .first()
    )


def create_employee(db: Session, employee: EmployeeCreate):
    db_employee = Employee(**employee.model_dump())

    db.add(db_employee)
    db.commit()
    db.refresh(db_employee)

    return db_employee


def update_employee(
    db: Session,
    employee_id: int,
    employee: EmployeeUpdate,
):
    db_employee = get_employee(db, employee_id)

    if not db_employee:
        return None

    update_data = employee.model_dump(exclude_unset=True)

    for key, value in update_data.items():
        setattr(db_employee, key, value)

    db.commit()
    db.refresh(db_employee)

    return db_employee


def delete_employee(db: Session, employee_id: int):
    db_employee = get_employee(db, employee_id)

    if not db_employee:
        return {"message":"Employee not found"}

    db.delete(db_employee)
    db.commit()

    return {"message":"Employee deleted"}